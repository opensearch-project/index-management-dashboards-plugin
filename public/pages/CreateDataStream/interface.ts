/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { RouteComponentProps } from "react-router-dom";
import { FieldInstance } from "../../lib/field";
import { DataStream } from "../../../server/models/interfaces";
import { TemplateItemRemote } from "../../../models/interfaces";

export interface DataStreamDetailProps {
  templateName?: string;
  onCancel?: () => void;
  onSubmitSuccess?: (templateName: string) => void;
  readonly?: boolean;
  history: RouteComponentProps["history"];
}

export interface SubDetailProps extends DataStreamDetailProps {
  field: FieldInstance;
  isEdit: boolean;
}

export interface DataStreamInEdit extends DataStream {
  matchedTemplate?: string;
}

export type TemplateItem = {
  name: string;
  index_template: TemplateItemRemote;
};
