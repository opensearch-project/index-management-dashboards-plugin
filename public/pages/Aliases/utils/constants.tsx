/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { SortDirection } from "../../../utils/constants";
import { IAlias } from "../interface";

export const DEFAULT_PAGE_SIZE_OPTIONS = [5, 10, 20, 50];

export const DEFAULT_QUERY_PARAMS = {
  from: "0",
  size: "20",
  search: "",
  sortField: "alias" as keyof IAlias,
  sortDirection: SortDirection.DESC,
  status: "",
  dataSourceId: "",
};
