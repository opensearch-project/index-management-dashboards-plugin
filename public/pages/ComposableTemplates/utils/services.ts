/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { CommonService } from "../../../services";
import { TemplateItemRemote } from "../../../../models/interfaces";

export const getAllUsedComponents = async ({ commonService }: { commonService: CommonService }) => {
  const allTemplatesResponse = await commonService.apiCaller<{
    index_templates?: Array<{
      name: string;
      index_template: TemplateItemRemote;
    }>;
  }>({
    endpoint: "transport.request",
    data: {
      method: "GET",
      path: "/_index_template/*",
    },
  });

  if (allTemplatesResponse && allTemplatesResponse.ok) {
    return allTemplatesResponse.response.index_templates || [];
  }

  return [];
};
