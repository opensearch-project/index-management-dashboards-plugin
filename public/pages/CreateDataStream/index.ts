/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import CreateDataStream from "./containers/CreateDataStream";

export default CreateDataStream;
