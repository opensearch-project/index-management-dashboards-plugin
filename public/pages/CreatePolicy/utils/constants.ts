/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const DEFAULT_POLICY = JSON.stringify(
  {
    policy: {
      description: "A simple default policy that changes the replica count between hot and cold states.",
      default_state: "example_hot_state",
      states: [
        {
          name: "example_hot_state",
          actions: [{ replica_count: { number_of_replicas: 5 } }],
          transitions: [{ state_name: "example_cold_state", conditions: { min_index_age: "30d" } }],
        },
        {
          name: "example_cold_state",
          actions: [{ replica_count: { number_of_replicas: 2 } }],
          transitions: [],
        },
      ],
      ism_template: {
        index_patterns: ["example-index-*"],
      },
    },
  },
  null,
  4
);
