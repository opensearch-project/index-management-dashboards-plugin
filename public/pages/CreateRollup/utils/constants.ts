/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const EMPTY_ROLLUP = JSON.stringify({
  rollup: {
    continuous: false,
    description: "",
    dimensions: [
      {
        date_histogram: {
          source_field: "",
          fixed_interval: "1h",
          timezone: "UTC",
        },
      },
    ],
    enabled: true,
    metrics: [],
    page_size: 1000,
    roles: [],
    schedule: {
      interval: {
        start_time: 234802,
        period: 1,
        unit: "MINUTES",
      },
    },
    source_index: "",
    target_index: "",
  },
});

export const DelayTimeunitOptions = [
  { value: "SECONDS", text: "Second(s)" },
  { value: "MINUTES", text: "Minute(s)" },
  { value: "HOURS", text: "Hour(s)" },
  { value: "DAYS", text: "Day(s)" },
];

export const ScheduleIntervalTimeunitOptions = [
  { value: "MINUTES", text: "Minute(s)" },
  { value: "HOURS", text: "Hour(s)" },
  { value: "DAYS", text: "Day(s)" },
];

export const AddFieldsColumns = [
  {
    field: "label",
    name: "Field name",
    sortable: true,
  },
  {
    field: "type",
    name: "Field type",
    sortable: true,
    render: (type: string | undefined) => (type == null || type === undefined ? "-" : type),
  },
];

export const WHERE_BOOLEAN_FILTERS = [
  { text: "Select value", value: "" },
  { text: "True", value: true },
  { text: "False", value: false },
];

export const OPERATORS_MAP = {
  IS: "is",
  IS_NOT: "is_not",
  IS_NULL: "is_null",
  IS_NOT_NULL: "is_not_null",
  IS_GREATER: "is_greater",
  IS_GREATER_EQUAL: "is_greater_equal",
  IS_LESS: "is_less",
  IS_LESS_EQUAL: "is_less_equal",
  STARTS_WITH: "starts_with",
  ENDS_WITH: "ends_with",
  CONTAINS: "contains",
  NOT_CONTAINS: "does_not_contains",
  IN_RANGE: "in_range",
  NOT_IN_RANGE: "not_in_range",
};

export const COMPARISON_OPERATORS = [
  { text: "is", value: OPERATORS_MAP.IS, dataTypes: ["number", "text", "keyword", "boolean"] },
  {
    text: "is not",
    value: OPERATORS_MAP.IS_NOT,
    dataTypes: ["number", "text", "keyword", "boolean"],
  },
  {
    text: "is null",
    value: OPERATORS_MAP.IS_NULL,
    dataTypes: ["number", "text", "keyword", "boolean"],
  },
  {
    text: "is not null",
    value: OPERATORS_MAP.IS_NOT_NULL,
    dataTypes: ["number", "text", "keyword"],
  },
  { text: "is greater than", value: OPERATORS_MAP.IS_GREATER, dataTypes: ["number"] },
  { text: "is greater than equal", value: OPERATORS_MAP.IS_GREATER_EQUAL, dataTypes: ["number"] },
  { text: "is less than", value: OPERATORS_MAP.IS_LESS, dataTypes: ["number"] },
  { text: "is less than equal", value: OPERATORS_MAP.IS_LESS_EQUAL, dataTypes: ["number"] },
  { text: "is in range", value: OPERATORS_MAP.IN_RANGE, dataTypes: ["number"] },
  { text: "is not in range", value: OPERATORS_MAP.NOT_IN_RANGE, dataTypes: ["number"] },
  { text: "starts with", value: OPERATORS_MAP.STARTS_WITH, dataTypes: ["text", "keyword"] },
  { text: "ends with", value: OPERATORS_MAP.ENDS_WITH, dataTypes: ["text", "keyword"] },
  { text: "contains", value: OPERATORS_MAP.CONTAINS, dataTypes: ["text", "keyword"] },
  { text: "does not contains", value: OPERATORS_MAP.NOT_CONTAINS, dataTypes: ["text"] },
];
