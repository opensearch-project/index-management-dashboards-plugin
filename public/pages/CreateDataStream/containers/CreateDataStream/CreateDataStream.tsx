/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Component } from "react";
import { RouteComponentProps } from "react-router-dom";
import { isEqual } from "lodash";
import DataStreamDetail from "../DataStreamDetail";
import { BREADCRUMBS, ROUTES } from "../../../../utils/constants";
import { CoreServicesContext } from "../../../../components/core_services";

type CreateDataStreamProps = RouteComponentProps<{ dataStream?: string }>;

export default class CreateDataStream extends Component<CreateDataStreamProps> {
  static contextType = CoreServicesContext;

  public get dataStream() {
    return this.props.match.params.dataStream;
  }

  setBreadCrumb() {
    const isEdit = this.dataStream;
    let lastBread: typeof BREADCRUMBS.TEMPLATES;
    if (isEdit) {
      lastBread = {
        text: this.dataStream || "",
        href: `#${this.props.location.pathname}`,
      };
    } else {
      lastBread = BREADCRUMBS.CREATE_DATA_STREAM;
    }
    this.context.chrome.setBreadcrumbs([BREADCRUMBS.INDEX_MANAGEMENT, BREADCRUMBS.DATA_STREAMS, lastBread]);
  }

  componentDidUpdate(prevProps: Readonly<CreateDataStreamProps>): void {
    if (!isEqual(prevProps, this.props)) {
      this.setBreadCrumb();
    }
  }

  componentDidMount = async (): Promise<void> => {
    this.setBreadCrumb();
  };

  onCancel = (): void => {
    this.props.history.push(ROUTES.DATA_STREAMS);
  };

  render() {
    return (
      <div style={{ padding: "0px 50px" }}>
        <DataStreamDetail
          history={this.props.history}
          dataStream={this.dataStream}
          onCancel={this.onCancel}
          onSubmitSuccess={() => this.props.history.push(ROUTES.DATA_STREAMS)}
        />
      </div>
    );
  }
}
