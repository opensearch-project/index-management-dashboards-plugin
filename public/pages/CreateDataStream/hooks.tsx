import { EuiFormRowProps } from "@elastic/eui";
import { flatten } from "flat";
import { get, set } from "lodash";
import { FieldInstance, transformNameToString } from "../../lib/field";
import { TemplateItem } from "./interface";
import { transformObjectToArray } from "../../components/IndexMapping";
import { filterByMinimatch } from "../../../utils/helper";

export const getCommonFormRowProps = (name: string | string[], field: FieldInstance): Partial<EuiFormRowProps> => {
  return {
    isInvalid: !!field.getError(name),
    error: field.getError(name),
    "data-test-subj": `form-row-${transformNameToString(name)}`,
  };
};

export const setMatchedTemplate = ({ matchedTemplate, field }: { matchedTemplate: TemplateItem; field: FieldInstance }) => {
  const payload = {
    matchedTemplate: matchedTemplate.name,
    template: JSON.parse(
      JSON.stringify({
        ...matchedTemplate.index_template.template,
        settings: flatten(matchedTemplate.index_template.template.settings || {}),
      })
    ),
  };

  set(payload, "template.mappings.properties", transformObjectToArray(get(payload, "template.mappings.properties", {})));

  field.setValues(payload);
};

export const getStringBeforeStar = (string: string) => string.replace(/^([^*]*).*$/, "$1");

export const findPatternMatchesString = (value: string, template: TemplateItem) => {
  const { index_patterns } = template.index_template;
  // matched by wildcard
  const findMatchesPattern = index_patterns.find((item) => filterByMinimatch(value, [item]));
  // matched by letters
  const findWordLikePattern = index_patterns.find((item) => item.match(new RegExp(value, "i")));
  // exactly the same
  const findStringPattern = index_patterns.find((item) => item === value);

  return {
    findMatchesPattern,
    findWordLikePattern,
    findStringPattern,
  };
};
