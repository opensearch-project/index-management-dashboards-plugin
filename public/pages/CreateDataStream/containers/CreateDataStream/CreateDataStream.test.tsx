/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { MemoryRouter as Router, Route, RouteComponentProps, Switch } from "react-router-dom";
import { render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import CreateDataStream from "./CreateDataStream";
import { ServicesContext } from "../../../../services";
import { browserServicesMock, coreServicesMock, apiCallerMock } from "../../../../../test/mocks";
import { ROUTES } from "../../../../utils/constants";
import { CoreServicesContext } from "../../../../components/core_services";

function renderCreateDataStreamWithRouter(initialEntries = [ROUTES.DATA_STREAMS] as string[]) {
  return {
    ...render(
      <Router initialEntries={initialEntries}>
        <CoreServicesContext.Provider value={coreServicesMock}>
          <ServicesContext.Provider value={browserServicesMock}>
            <Switch>
              <Route
                path={`${ROUTES.CREATE_DATA_STREAM}/:template`}
                render={(props: RouteComponentProps) => <CreateDataStream {...props} />}
              />
              <Route path={ROUTES.CREATE_DATA_STREAM} render={(props: RouteComponentProps) => <CreateDataStream {...props} />} />
              <Route path={ROUTES.DATA_STREAMS} render={(props: RouteComponentProps) => <h1>location is: {ROUTES.DATA_STREAMS}</h1>} />
            </Switch>
          </ServicesContext.Provider>
        </CoreServicesContext.Provider>
      </Router>
    ),
  };
}

describe("<CreateDataStream /> spec", () => {
  beforeEach(() => {
    apiCallerMock(browserServicesMock);
  });
  it("it goes to data streams page when click cancel", async () => {
    const { getByTestId, getByText, container, findByText } = renderCreateDataStreamWithRouter([ROUTES.CREATE_DATA_STREAM]);
    await findByText("Define data stream");
    expect(container).toMatchSnapshot();
    userEvent.click(getByTestId("CreateDataStreamCancelButton"));
    await waitFor(() => {
      expect(getByText(`location is: ${ROUTES.DATA_STREAMS}`)).toBeInTheDocument();
    });
  });
});
