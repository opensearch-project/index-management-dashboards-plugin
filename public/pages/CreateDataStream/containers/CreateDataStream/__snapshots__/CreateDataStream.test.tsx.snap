// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateDataStream /> spec it goes to data streams page when click cancel 1`] = `
<div>
  <div
    style="padding: 0px 50px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
    >
      <div
        class="euiFlexItem"
      >
        <h1
          class="euiTitle euiTitle--large"
        >
          Create data stream
        </h1>
        <div
          class="euiFormRow euiFormRow--fullWidth"
          id="some_html_id-row"
        >
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              class="euiFormHelpText euiFormRow__text"
              style="padding-top: 0px; padding-bottom: 4px;"
            >
              <div>
                Data streams simplify the management of time-series data. Data streams are composed of multiple backing indices. Search requests are routed to all backing indexes, while indexing requests are routed to the latest write index.
                 
                <button
                  class="euiLink euiLink--primary"
                  type="button"
                >
                  Learn more.
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        style="padding: 0px 10px;"
      >
        <div
          class="euiFlexItem"
        >
          <h3
            class="euiTitle euiTitle--small"
          >
            Define data stream
            <span
              class="panel-header-count"
            >
               
               
            </span>
          </h3>
        </div>
      </div>
      <hr
        class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
      />
      <div
        style="padding: 0px 10px;"
      >
        <div
          class="euiSpacer euiSpacer--s"
        />
        <div
          class="euiFormRow euiFormRow--fullWidth"
          data-test-subj="form-row-name"
          id="some_html_id-row"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              aria-invalid="false"
              class="euiFormLabel euiFormRow__label"
              for="some_html_id"
            >
              Data stream name
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              class="euiFormHelpText euiFormRow__text"
              style="padding-top: 0px; padding-bottom: 4px;"
            >
              Enter a data stream name. It must match an index pattern from a data stream template.
               
              <a
                class="euiLink euiLink--primary"
                href="#/templates"
                rel="noopener noreferrer"
                target="_blank"
              >
                Manage templates
                EuiIconMock
                <span
                  class="euiScreenReaderOnly"
                >
                  (opens in a new tab or window)
                </span>
              </a>
            </div>
            <span
              class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                class="euiComboBox"
                role="combobox"
              >
                <div
                  class="euiFormControlLayout"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <div
                      class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isClearable"
                      data-test-subj="comboBoxInput"
                      tabindex="-1"
                    >
                      <p
                        class="euiComboBoxPlaceholder"
                      >
                        Specify data stream name
                      </p>
                      <div
                        class="euiComboBox__input"
                        style="font-size: 14px; display: inline-block;"
                      >
                        <input
                          aria-controls=""
                          data-test-subj="comboBoxSearchInput"
                          role="textbox"
                          style="box-sizing: content-box; width: 2px;"
                          value=""
                        />
                        <div
                          style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                        />
                      </div>
                    </div>
                    <div
                      class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                    >
                      <button
                        aria-label="Open list of options"
                        class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                        data-test-subj="comboBoxToggleListButton"
                        type="button"
                      >
                        EuiIconMock
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </span>
          </div>
        </div>
        <div
          class="euiSpacer euiSpacer--l"
        />
        <div
          class="euiFormRow"
          id="some_html_id-row"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              class="euiFormLabel euiFormRow__label"
              for="some_html_id"
            >
              Matching index template
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            No index template matched
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
    >
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <button
          class="euiButtonEmpty euiButtonEmpty--primary"
          data-test-subj="CreateDataStreamCancelButton"
          type="button"
        >
          <span
            class="euiButtonContent euiButtonEmpty__content"
          >
            <span
              class="euiButtonEmpty__text"
            >
              Cancel
            </span>
          </span>
        </button>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <button
          class="euiButton euiButton--primary euiButton--fill"
          data-test-subj="CreateDataStreamCreateButton"
          type="button"
        >
          <span
            class="euiButtonContent euiButton__content"
          >
            <span
              class="euiButton__text"
            >
              Create data stream
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
